spartan_parallel/src/scalar/fp.rs

```diff
@@ -0,0 +1,204 @@
+use super::SpartanExtensionField;
+use crate::{AppendToTranscript, ProofTranscript, Transcript};
+use ceno_goldilocks::Goldilocks;
+use core::borrow::Borrow;
+use core::iter::{Product, Sum};
+use core::ops::{Add, AddAssign, Mul, MulAssign, Sub, SubAssign};
+use ff::{Field, FromUniformBytes};
+use rand::{CryptoRng, RngCore};
+use serde::{Deserialize, Serialize};
+use std::ops::Neg;
+use subtle::{Choice, ConditionallySelectable, ConstantTimeEq, CtOption};
+use zeroize::Zeroize;
+
+/// Constant representing the modulus
+/// q = 2^64 - 2^32 + 1
+/// 0xFFFFFFFF00000001
+const P: u64 = 0xFFFF_FFFF_0000_0001;
+
+/// Field wrapper around base Goldilocks
+#[derive(Clone, Copy, Eq, Serialize, Deserialize, Hash, Debug)]
+pub struct Scalar(Goldilocks);
+
+impl SpartanExtensionField for Scalar {
+  type InnerType = Goldilocks;
+
+  fn inner(&self) -> &Goldilocks {
+    &self.0
+  }
+
+  fn field_zero() -> Self {
+    Goldilocks::ZERO.into()
+  }
+
+  fn field_one() -> Self {
+    Goldilocks::ONE.into()
+  }
+
+  fn random<Rng: RngCore + CryptoRng>(rng: &mut Rng) -> Self {
+    let mut res = rng.next_u64();
+    while res >= P {
+      res = rng.next_u64();
+    }
+    Goldilocks(res).into()
+  }
+
+  /// Attempts to convert a little-endian byte representation of
+  /// a scalar into a `Scalar`, failing if the input is not canonical.
+  fn from_bytes(bytes: &[u8; 32]) -> CtOption<Scalar> {
+    CtOption::new(
+      Goldilocks::from_uniform_bytes(bytes).into(),
+      Choice::from(1u8),
+    )
+  }
+
+  /// Converts an element of `Scalar` into a byte representation in
+  /// little-endian byte order.
+  fn to_bytes(&self) -> [u8; 32] {
+    let mut res = [0; 32];
+    res[..8].copy_from_slice(&self.0 .0.to_le_bytes());
+    res
+  }
+
+  /// Converts a 512-bit little endian integer into
+  /// a `Scalar` by reducing by the modulus.
+  fn from_bytes_wide(bytes: &[u8; 64]) -> Scalar {
+    Goldilocks::from_uniform_bytes(bytes).into()
+  }
+
+  /// Append Goldilocks scalar to transcript
+  fn append_field_to_transcript(label: &'static [u8], transcript: &mut Transcript, input: Self) {
+    transcript.append_scalar(label, &input);
+  }
+
+  /// Append a vector Goldilocks scalars to transcript
+  fn append_field_vector_to_transcript(
+    label: &'static [u8],
+    transcript: &mut Transcript,
+    input: &[Self],
+  ) {
+    transcript.append_message(label, b"begin_append_vector");
```

Comment:
Why do we need these messages?  Who's supposed to read them?

Reply:
Ok. I'm just asking because it looked like new code.

You don't have to change the original code base here, the PR is already but enough. Thanks.

Reply:
Ah ok. Marked as resolved